client/order/orderbook.go

```diff
@@ -0,0 +1,226 @@
+
+	return nil
+}
+
+// Initialize synchronizes a client tracked order book with an order
+// book snapshot.
+func (ob *OrderBook) Initialize(snapshot *msgjson.OrderBook) error {
+	ob.mtx.Lock()
+	defer ob.mtx.Unlock()
+
+	if ob.synced {
+		return fmt.Errorf("reinitializing a synced order book")
+	}
+
+	ob.orders = make(map[string]*msgjson.BookOrderNote, len(snapshot.Orders))
+	ob.buys = make(map[uint64][]*msgjson.BookOrderNote)
+	ob.sells = make(map[uint64][]*msgjson.BookOrderNote)
+
+	for _, order := range snapshot.Orders {
+		ob.addOrder(order)
+	}
+
+	ob.seq = snapshot.Seq
+	ob.marketID = snapshot.MarketID
+	ob.synced = true
+
+	return nil
+}
+
+// Book adds a new order to the order book.
+func (ob *OrderBook) Book(order *msgjson.BookOrderNote) error {
+	ob.mtx.Lock()
+	defer ob.mtx.Unlock()
+
+	if !ob.synced {
+		return fmt.Errorf("order book is unsynced")
+	}
+
+	if ob.marketID != order.MarketID {
+		return fmt.Errorf("market id mismatch, expected %s got %s",
+			ob.marketID, order.MarketID)
+	}
+
+	expectedSeq := ob.seq + 1
+	if expectedSeq != order.Seq {
+		ob.synced = false
+		return fmt.Errorf("order book out of sync, expected %d got %d",
+			expectedSeq, order.Seq)
+	}
+
+	ob.seq++
+	ob.addOrder(order)
+	return nil
+}
+
+// Unbook removes an order from the order book.
+func (ob *OrderBook) Unbook(unbookOrder *msgjson.UnbookOrderNote) error {
+	ob.mtx.Lock()
+	defer ob.mtx.Unlock()
+
+	if !ob.synced {
+		return fmt.Errorf("order book is unsynced")
+	}
+
+	if ob.marketID != unbookOrder.MarketID {
+		return fmt.Errorf("market id mismatch, expected %s got %s",
+			ob.marketID, unbookOrder.MarketID)
+	}
+
+	order, ok := ob.orders[unbookOrder.OrderID.String()]
+	if !ok {
+		return fmt.Errorf("no order found with id %v",
+			unbookOrder.OrderID.String())
+	}
+
+	expectedSeq := ob.seq + 1
+	if expectedSeq != order.Seq {
+		ob.synced = false
+		return fmt.Errorf("order book out of sync, expected %v got %v",
+			expectedSeq, order.Seq)
+	}
+
+	ob.seq++
+	err := ob.removeOrder(order)
+	if err != nil {
+		ob.synced = false
+		return err
+	}
+	return nil
+}
+
+// Synced returns the sync state of the order book.
+func (ob *OrderBook) Synced() bool {
+	ob.mtx.RLock()
+	defer ob.mtx.RUnlock()
+	return ob.synced
+}
+
+// HighestBuyPrice returns the highest buying price from the order book.
+func (ob *OrderBook) HighestBuyPrice() (uint64, error) {
+	ob.mtx.RLock()
+	defer ob.mtx.RUnlock()
+
+	if !ob.synced {
+		return 0, fmt.Errorf("order book is unsynced")
+	}
+
+	var highest uint64
+	for price := range ob.buys {
+		if highest < price {
+			highest = price
+		}
+	}
+
+	return highest, nil
+}
+
+// LowestSellPrice returns the lowest selling price from the order book.
+func (ob *OrderBook) LowestSellPrice() (uint64, error) {
+	ob.mtx.RLock()
+	defer ob.mtx.RUnlock()
+
+	if !ob.synced {
+		return 0, fmt.Errorf("order book is unsynced")
+	}
+
+	var lowest uint64
+	for price := range ob.sells {
+		if lowest > price {
+			lowest = price
+		}
+	}
```

Comment:
I'm less concerned with the performance cost here. Just to speculate wildly with no evidence, I'd guess that up to 1,000 orders would still be pretty zippy with simpler structures and sorts like this. Not sure about 10,000. 

If we did want to mirror the server's binary tree implementation, it looks to me like it would need to be written from scratch, but I'm not sure. Is it worth trying to generalize `OrderPQ` so that it can be shared?

Reply:
We can punt on this, but I expect a mature market with have several thousand orders.  But I don't know the rate... a couple per minute? Per second?

Reply:
I agree this isn't efficient, looking into sorting the initial orderbook and maintaining the order when new entries are added/removed. This should also help with fetching the best n order for a given quantity.